[content-visibility] Sync visibility bits in RenderLayer with content skipping

rwlbuis opened the pull request:
#### b9156186f52f48059654cca9eb179ce656cd86db
<pre>
[content-visibility] Sync visibility bits in RenderLayer with content skipping
<a href="https://bugs.webkit.org/show_bug.cgi?id=251614">https://bugs.webkit.org/show_bug.cgi?id=251614</a>

Reviewed by Simon Fraser.

Sync visibility bits in RenderLayer with content skipping by making sure
skipped content layers have m_hasVisibleContent and m_hasVisibleDescendant
set to false, and have m_hasVisibleDescendant set to false for layers
created by content-visibility: hidden (layer here is forced because of the
implicitly enabled layout containment).

This enables any hidden layer optimizations to apply to skipped content.

This change also removes some unneeded changes from r256186
and fixes the content-visibility-video.html WPT test.

* LayoutTests/TestExpectations:
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::RenderLayer):
(WebCore::RenderLayer::updateDescendantDependentFlags):
(WebCore::RenderLayer::computeHasVisibleContent const):
(WebCore::RenderLayer::paintList): undo old change
(WebCore::RenderLayer::calculateClipRects const):
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::updateAfterDescendants): undo old change
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::canBeComposited const):

Canonical link: <a href="https://commits.webkit.org/267547@main">https://commits.webkit.org/267547@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/8936862ade42a6d60a462c288f74f9881093d0e4

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/16926 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/17249 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/17708 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/18709 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/15845 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/17116 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/20481 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/17387 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/18091 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/17125 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/17483 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/14653 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/19518 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/14726 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/15348 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/22070 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/15710 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/15515 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/19830 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/16123 "Built successfully") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/17/builds/13660 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/15286 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/4054 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/19650 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/15953 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

smfr commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -370,6 +370,10 @@ RenderLayer::RenderLayer(RenderLayerModelObject& renderer)
 
+    //  We need the parent to know if we have skipped content or content-visibility root.
+    if (renderer.style().effectiveSkippedContent() && !renderer.parent())
+        return;
```
Not really a fan of these early returns from a constructor; easy to add code at the end and make mistakes.

rwlbuis replied:
One way around this is adding a boolean variable in the ctor to decide whether to do the hasVisibleContent. Let me know in case you would like me to do that in this patch or a follow-up.

smfr replied:
You could move code inside a lambda function, and just early return from that when m_visibleContentStatusDirty gets false. That's the cleanest.

rwlbuis replied:
Nice, I forgot about the lambda option, fixed.

smfr commented on Source/WebCore/rendering/RenderLayer.cpp:
```diff
@@ -1614,7 +1618,12 @@ void RenderLayer::updateDescendantDependentFlags()
 
-        for (RenderLayer* child = firstChild(); child; child = child->nextSibling()) {
+        auto firstLayerChild = [&] () -> RenderLayer* {
+            if (renderer().style().contentVisibility() != ContentVisibility::Visible)
```
What happens if we call `updateDescendantDependentFlags()` on a layer that's inside, but not the root of a contentVisibility subtree?

rwlbuis replied:
In that case the calculation proceeds as normal and since all the skipped content layers have m_hasVisibleContent set to false the calculation will result in hasVisibleDescendant remaining false for the root. So in a way it is doing unneeded work, OTOH adding another condition in the firstLayerChild lambda, while fixing the unneeded work, slows down the common case. I am not sure what option is better.

rwlbuis committed:
[content-visibility] Sync visibility bits in RenderLayer with content skipping
https://bugs.webkit.org/show_bug.cgi?id=251614

Reviewed by Simon Fraser.

Sync visibility bits in RenderLayer with content skipping by making sure
skipped content layers have m_hasVisibleContent and m_hasVisibleDescendant
set to false, and have m_hasVisibleDescendant set to false for layers
created by content-visibility: hidden (layer here is forced because of the
implicitly enabled layout containment).

This enables any hidden layer optimizations to apply to skipped content.

This change also removes some unneeded changes from r256186
and fixes the content-visibility-video.html WPT test.

* LayoutTests/TestExpectations:
* Source/WebCore/rendering/RenderLayer.cpp:
(WebCore::RenderLayer::RenderLayer):
(WebCore::RenderLayer::updateDescendantDependentFlags):
(WebCore::RenderLayer::computeHasVisibleContent const):
(WebCore::RenderLayer::paintList): undo old change
(WebCore::RenderLayer::calculateClipRects const):
* Source/WebCore/rendering/RenderLayerBacking.cpp:
(WebCore::RenderLayerBacking::updateAfterDescendants): undo old change
* Source/WebCore/rendering/RenderLayerCompositor.cpp:
(WebCore::RenderLayerCompositor::canBeComposited const):

Canonical link: https://commits.webkit.org/267547@main
